server/src-lib/Hasura/RQL/Types/Catalog.hs

```diff
@@ -139,6 +140,20 @@ instance NFData CatalogFunction
 instance Cacheable CatalogFunction
 $(deriveFromJSON (aesonDrop 3 snakeCase) ''CatalogFunction)
 
+data CatalogCustomTypes
+  = CatalogCustomTypes
+  { _cctCustomTypes :: !CustomTypes
+  , _cctPgScalars   :: ![PGScalarType]
+  -- ^ Base types from Postgres required for validating custom types
+  -- to check if any one of them uses the Postgres base type name
+  -- and publish them in generated GraphQL schema as Scalars.
+  -- These won't be tracked in the metadata.
+  -- See https://github.com/hasura/graphql-engine/issues/4125
```

Comment:
> `Base types from Postgres required for validating custom types to check if any one of them uses the Postgres base type name and publish them in generated GraphQL schema as Scalars.`

I’m afraid I don’t understand this sentence. 🙁 Perhaps it could be reworded as something like this:

```haskell
-- ^ All Postgres base types, which may be referenced in custom type
-- definitions. When we validate the custom types (see
-- 'validateCustomTypeDefinitions'), we record which base types were referenced
-- so that we can be sure to include them in the generated GraphQL schema.
```

> `These won't be tracked in the metadata.`

Why not? Presumably because we fetch them on-demand from the types in the GraphQL database, but then it would be good to explicitly say so!

```haskell
-- These are not actually part of the Hasura metadata --- we fetch them from
-- @pg_catalog.pg_type@ --- but they’re needed when validating the custom type
-- metadata, so we include them here.
```

> `See https://github.com/hasura/graphql-engine/issues/4125`

Generally, I like linking relevant issues! But in this case, #4125 doesn’t really include any useful information in its description (which is unfortunate, but a separate problem), so I’d just drop the link.